File: hw/src/main/java/com/sumian/sleepdoctor/base/BaseWebViewActivity.java
package com.sumian.sleepdoctor.base;

import android.arch.lifecycle.LifecycleOwner;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebView;

import com.blankj.utilcode.util.LogUtils;
import com.sumian.sleepdoctor.BuildConfig;
import com.sumian.sleepdoctor.R;
import com.sumian.sleepdoctor.app.AppManager;
import com.sumian.sleepdoctor.h5.bean.H5ShowToastData;
import com.sumian.sleepdoctor.utils.ScreenUtil;
import com.sumian.sleepdoctor.utils.SoftKeyBoardListener;
import com.sumian.sleepdoctor.widget.TitleBar;
import com.sumian.sleepdoctor.widget.dialog.SumianImageTextDialog;
import com.sumian.sleepdoctor.widget.webview.SBridgeHandler;
import com.sumian.sleepdoctor.widget.webview.SWebView;
import com.sumian.sleepdoctor.widget.webview.SWebViewLayout;

import butterknife.BindView;

/**
 * Created by sm
 * on 2018/5/25 10:03
 * desc:
 **/
public abstract class BaseWebViewActivity<Presenter extends BasePresenter> extends BaseActivity<Presenter> implements TitleBar.OnBackClickListener, SWebViewLayout.WebListener {

    @BindView(R.id.sm_webview_container)
    protected SWebViewLayout mSWebViewLayout;
    @BindView(R.id.title_bar)
    TitleBar mTitleBar;
    @BindView(R.id.root_view)
    View mRootView;
    private SoftKeyBoardListener mSoftKeyBoardListener;
    private SumianImageTextDialog mSumianImageTextDialog;

    @Override
    protected int getLayoutId() {
        return R.layout.activity_main_base_webview;
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        monitorKeyboard();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        mSoftKeyBoardListener.release();
    }

    @Override
    protected void initWidget(View root) {
        super.initWidget(root);
        mTitleBar.setOnBackClickListener(this);
    }

    @Override
    protected void initData() {
        super.initData();
        SWebView sWebView = mSWebViewLayout.getSWebView();
        mSWebViewLayout.setWebListener(this);
        mSWebViewLayout.loadRequestUrl(getCompleteUrl());
        registerHandler(sWebView);
        registerBaseHandler(sWebView);
    }

    private void registerBaseHandler(SWebView sWebView) {
        sWebView.registerHandler("showToast", new SBridgeHandler() {
            @SuppressWarnings("ConstantConditions")
            @Override
            public void handler(String data) {
                LogUtils.d(data);
                H5ShowToastData toastData = H5ShowToastData.Companion.fromJson(data);
                if (mSumianImageTextDialog != null) {
                    mSumianImageTextDialog.dismiss();
                } else {
                    mSumianImageTextDialog = new SumianImageTextDialog(mActivity);
                }
                mSumianImageTextDialog.show(toastData);
            }
        });
        sWebView.registerHandler("hideToast", new SBridgeHandler() {
            @SuppressWarnings("ConstantConditions")
            @Override
            public void handler(String data) {
                LogUtils.d(data);
                H5ShowToastData toastData = H5ShowToastData.Companion.fromJson(data);
                if (mSumianImageTextDialog != null) {
                    mSumianImageTextDialog.dismiss(toastData.getDelay());
                }
            }
        });
        sWebView.registerHandler("finish", new SBridgeHandler() {
            @Override
            public void handler(String data) {
                finish();
            }
        });
    }


    @Override
    public void onStart(@NonNull LifecycleOwner owner) {
        super.onStart(owner);
    }

    @Override
    protected void onResume() {
        super.onResume();
        mSWebViewLayout.resumeWebView();
    }

    @Override
    protected void onPause() {
        super.onPause();
        mSWebViewLayout.pauseWebView();
    }

    @Override
    protected void onRelease() {
        super.onRelease();
        mSWebViewLayout.destroyWebView();
    }

    @Override
    protected void onStop() {
        if (mSumianImageTextDialog != null) {
            mSumianImageTextDialog.release();
        }
        super.onStop();
    }

    protected String h5HandlerName() {
        return null;
    }

    protected String initTitle() {
        return null;
    }

    protected void registerHandler(@NonNull SWebView sWebView) {
    }

    @Override
    public void onBack(View v) {
        onBackPressed();
    }

    @Override
    public void onBackPressed() {
        if (!mSWebViewLayout.webViewCanGoBack()) {
            super.onBackPressed();
        }
    }

    protected String getCompleteUrl() {
        StringBuilder stringBuilder = new StringBuilder();
        String urlServerPart = getUrlServerPart();
        String urlContentPart = getUrlContentPart();
        boolean contentContainerParams = urlContentPart.contains("?");
        stringBuilder.append(urlServerPart)
                .append(urlContentPart)
                .append(contentContainerParams ? "&" : "?")
                .append(getUrlToken());
        String url = stringBuilder.toString();
        LogUtils.d("load url: %s", url);
        return url;
    }

    private String getUrlServerPart() {
        return BuildConfig.BASE_H5_URL;
    }

    protected String getUrlContentPart() {
        return null;
    }

    private String getUrlToken() {
        return "token=" + AppManager.getAccountViewModel().accessToken();
    }

    @Override
    public void onReceiveTitle(WebView webView, String title) {
        mTitleBar.setTitle(title);
    }

    private void monitorKeyboard() {
        mSoftKeyBoardListener = SoftKeyBoardListener.registerListener(this, new SoftKeyBoardListener.OnSoftKeyBoardChangeListener() {
            @Override
            public void keyBoardShow(int height) {
                updateRootViewHeight(ScreenUtil.getScreenHeight(mActivity) - height);
            }

            @Override
            public void keyBoardHide(int height) {
                updateRootViewHeight(ScreenUtil.getScreenHeight(mActivity));
            }
        });
    }

    private void updateRootViewHeight(int height) {
        ViewGroup.LayoutParams layoutParams = mRootView.getLayoutParams();
        layoutParams.height = height;
        mRootView.requestLayout();
    }

    protected void reload() {
        mSWebViewLayout.reload();
    }
}

File: common/src/main/java/com/sumian/common/h5/BaseWebViewActivity.kt
package com.sumian.common.h5

import android.os.Bundle
import android.view.View
import com.blankj.utilcode.util.LogUtils
import com.github.lzyzsd.jsbridge.BridgeHandler
import com.google.gson.reflect.TypeToken
import com.sumian.common.R
import com.sumian.common.base.BasePresenterActivity
import com.sumian.common.dialog.SumianImageTextDialog
import com.sumian.common.h5.bean.H5ShowToastData
import com.sumian.common.h5.bean.NativeRouteData
import com.sumian.common.h5.bean.ShareData
import com.sumian.common.h5.widget.SWebView
import com.sumian.common.mvp.IPresenter
import com.sumian.common.utils.JsonUtil
import com.sumian.common.utils.ScreenUtil
import com.sumian.common.utils.StatusBarUtil
import com.sumian.common.widget.TitleBar
import com.tencent.smtt.sdk.WebView
import kotlinx.android.synthetic.main.common_activity_base_webview.*

/**
 * <pre>
 *     @author : Zhan Xuzhao
 *     time   : 2018/9/10 10:53
 *     desc   :
 *     version: 1.0
 * </pre>
 */
@Suppress("MemberVisibilityCanBePrivate", "UNUSED_ANONYMOUS_PARAMETER")
abstract class BaseWebViewActivity : BasePresenterActivity<IPresenter>(), SWebView.OnWebViewListener {

    protected var mSoftKeyBoardListener: SoftKeyBoardListener? = null
    protected var mSumianImageTextDialog: SumianImageTextDialog? = null

    override fun getLayoutId(): Int {
        return R.layout.common_activity_base_webview
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        monitorKeyboard()
    }

    override fun onDestroy() {
        super.onDestroy()
        mSoftKeyBoardListener?.release()
    }

    override fun showBackNav(): Boolean {
        return true
    }

    override fun initData() {
        super.initData()
        val sWebView = sm_webview_container.sWebView
        sm_webview_container?.setWebListener(this)
        sm_webview_container?.loadRequestUrl(getCompleteUrl())
        registerHandler(sWebView)
        registerBaseHandler(sWebView)
    }

    private fun registerBaseHandler(sWebView: SWebView) {
        sWebView.registerHandler("showToast") { data, function ->
            LogUtils.d(data)
            val toastData = H5ShowToastData.fromJson(data)
            if (mSumianImageTextDialog != null) {
                mSumianImageTextDialog!!.dismiss()
            } else {
                mSumianImageTextDialog = SumianImageTextDialog(this@BaseWebViewActivity)
            }
            mSumianImageTextDialog!!.show(toastData)
        }
        sWebView.registerHandler("hideToast") { data, function ->
            LogUtils.d(data)
            val toastData = H5ShowToastData.fromJson(data)
            if (mSumianImageTextDialog != null) {
                mSumianImageTextDialog!!.dismiss(toastData.delay)
            }
        }
        sWebView.registerHandler("finish") { data, function -> finish() }
        sWebView.registerHandler("return") { data, function -> onBackPressed() }
        sWebView.registerHandler("updatePageUI", BridgeHandler { data, function ->
            val map = JsonUtil.fromJson<Map<String, Any>>(data, object : TypeToken<Map<String, Any>>() {

            }.type) ?: return@BridgeHandler
            for ((key, value) in map) {
                when (key) {
                    "showNavigationBar" -> if (value is Boolean) {
                        mTitleBar.visibility = if (value) View.VISIBLE else View.GONE
                    }
                    "showTitle" -> if (value is Boolean) {
                        mTitleBar.showTitle(value)
                    }
                    "showBackArrow" -> if (value is Boolean) {
                        mTitleBar.showBackArrow(value)
                    }
                    "setStatusBarTextColorDark" -> if (value is Boolean) {
                        StatusBarUtil.setStatusBarTextColorDark(this@BaseWebViewActivity, value)
                    }
                    else -> {
                    }
                }
            }
        })
        sWebView.registerHandler("goToPage") { data, function ->
            run {
                val routeData = JsonUtil.fromJson(data, NativeRouteData::class.java) ?: return@run
                onGoToPage(routeData)
            }
        }
        sWebView.registerHandler("share") { data, function ->
            run {
                val shareData = JsonUtil.fromJson(data, ShareData::class.java) ?: return@run
                onShare(shareData)
            }
        }
    }

    protected open fun onGoToPage(routeData: NativeRouteData) {

    }

    protected open fun onShare(shareData: ShareData) {

    }

    override fun onResume() {
        super.onResume()
        sm_webview_container!!.resumeWebView()
    }

    override fun onPause() {
        super.onPause()
        sm_webview_container!!.pauseWebView()
    }

    override fun onRelease() {
        super.onRelease()
        sm_webview_container!!.destroyWebView()
    }

    override fun onStop() {
        mSumianImageTextDialog?.release()
        super.onStop()
    }

    protected open fun h5HandlerName(): String? {
        return null
    }

    protected open fun initTitle(): String? {
        return null
    }

    protected open fun registerHandler(sWebView: SWebView) {}

    override fun onBackPressed() {
        if (!sm_webview_container!!.webViewCanGoBack()) {
            super.onBackPressed()
        }
    }

    protected open fun getCompleteUrl(): String {
        val stringBuilder = StringBuilder()
        val urlServerPart = getUrlServerPart()
        val urlContentPart = getUrlContentPart()
        val contentContainerParams = urlContentPart != null && urlContentPart.contains("?")
        stringBuilder.append(urlServerPart)
                .append(urlContentPart)
                .append(if (contentContainerParams) "&" else "?")
                .append(getUrlToken())
        val url = stringBuilder.toString()
        LogUtils.d("load url: %s", url)
        return url
    }

    protected fun getUrlServerPart(): String {
        return WebViewManger.getInstance().getBaseUrl() ?: ""
    }

    protected open fun getUrlContentPart(): String? {
        return null
    }

    private fun getUrlToken(): String {
        return "token=" + getToken()
    }

    protected fun getToken(): String? {
        return WebViewManger.getInstance().getToken()
    }


    protected open fun monitorKeyboard() {
        mSoftKeyBoardListener = SoftKeyBoardListener.registerListener(this, object : SoftKeyBoardListener.OnSoftKeyBoardChangeListener {
            override fun keyBoardShow(height: Int) {
                updateRootViewHeight(ScreenUtil.getScreenHeight(this@BaseWebViewActivity) - height)
            }

            override fun keyBoardHide(height: Int) {
                updateRootViewHeight(ScreenUtil.getScreenHeight(this@BaseWebViewActivity))
            }
        })
    }

    private fun updateRootViewHeight(height: Int) {
        val layoutParams = mBackRootView.layoutParams
        layoutParams.height = height
        mBackRootView.requestLayout()
    }

    protected fun reload() {
        sm_webview_container!!.reload()
    }

    protected fun getTitleBar(): TitleBar {
        return mTitleBar
    }

    protected fun getWebView(): SWebView {
        return sm_webview_container.sWebView
    }

    // ----------- OnWebViewListener start -----------
    protected fun canWebGoBack(): Boolean {
        return getWebView().canGoBack()
    }

    override fun onPageStarted(view: WebView?) {
    }

    override fun onProgressChange(view: WebView?, newProgress: Int) {
    }

    override fun onPageFinish(view: WebView?) {
    }

    override fun onRequestErrorCallback(view: WebView?, responseCode: Int) {
    }

    override fun onRequestNetworkErrorCallback(view: WebView?) {
    }

    override fun onReceiveTitle(webView: WebView, title: String) {
        mTitleBar.setTitle(title)
    }
    // ----------- OnWebViewListener end -----------
}